### PR TITLE
Add replacement headers for std thread, mutex and condition_variable

### PR DESCRIPTION
**Pull request by josephnoir**
I would like to contribute header files that allow functionality similar to the STL thread, mutex and condition_variable. For now, I have added these files to sys as a module and included some basic tests.

Any thoughts or comments?

References: [thread](http://en.cppreference.com/w/cpp/thread/thread), [mutex](http://en.cppreference.com/w/cpp/thread/mutex), [unique_lock](http://en.cppreference.com/w/cpp/thread/unique_lock), [condition_variable](http://en.cppreference.com/w/cpp/thread/condition_variable)

### CODE REVIEW

**Comment by BytesGalore** [comment]
@josephnoir Woohoo std::thread, lambda functions, locks and template programming :smile: 
Awesome RIOT becomes C++ified :+1:

File: sys/stl-compatibility/condition_variable.cpp [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+void condition_variable::wait(unique_lock<mutex>& lock) noexcept {
+  if (!lock.owns_lock()) {
+    throw std::runtime_error("Mutex not locked.");
+  }
+  priority_queue_node_t n;
+  n.priority = sched_active_thread->priority;
+  n.data = sched_active_pid;
+  n.next = NULL;
+  // the signaling thread may not hold the mutex, the queue is not thread safe
```
Comment (BytesGalore) [inline sys/stl-compatibility/condition_variable.cpp]: please change the comments form `//` style to `/* ... */`

File: sys/stl-compatibility/condition_variable.cpp [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  }
+  restoreIRQ(old_state);
+  if (other_prio >= 0) {
+    sched_switch(other_prio);
+  }
+}
+
+void condition_variable::wait(unique_lock<mutex>& lock) noexcept {
+  if (!lock.owns_lock()) {
+    throw std::runtime_error("Mutex not locked.");
```
Comment (BytesGalore) [inline sys/stl-compatibility/condition_variable.cpp]: seems to be the only explicit `throw`, does it still hardfaults a board?

File: sys/stl-compatibility/include/riot/thread.hpp [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+ public:
+  using id = thread_id;
+  using native_handle_type = kernel_pid_t;
+
+  inline thread() noexcept : m_handle{thread_uninitialized} {}
+  template <class F, class... Args,
+            class = typename std::enable_if
+            <!std::is_same<typename std::decay<F>::type, thread>::value>::type>
+  explicit thread(F&& f, Args&&... args);
```
Comment (BytesGalore) [inline sys/stl-compatibility/include/riot/thread.hpp]: phew, I feel that I will need to consult you for some tutoring @josephnoir

File: sys/stl-compatibility/condition_variable.cpp (hunk 1 above)
Comment (gebart) [inline sys/stl-compatibility/condition_variable.cpp]: Does this code convention still apply in C++ code?

Comment (josephnoir) [inline sys/stl-compatibility/condition_variable.cpp]: I would argue to reserve ```/* ... */``` function/class documentation, headers and commenting out blocks while coding. 

None the less, I can change this to comply with the coding style. When are you allowed to use ```//```?

Comment (OlegHahm) [inline sys/stl-compatibility/condition_variable.cpp]: > Does this code convention still apply to C++ code?

Valid question. I would say no, but we'll have to update the conventions then.

File: sys/stl-compatibility/condition_variable.cpp (hunk 2 above)
Comment (josephnoir) [inline sys/stl-compatibility/condition_variable.cpp]: There are more functions that throw exceptions (as specified by the reference). However this is the only runtime error, I believe the others are system_errors as this one should be. Let me change that ...

File: sys/stl-compatibility/include/riot/thread.hpp (hunk 3 above)
Comment (josephnoir) [inline sys/stl-compatibility/include/riot/thread.hpp]: This is not required by the reference. I will slim it down accordingly ^^

File: sys/stl-compatibility/condition_variable.cpp (hunk 1 above)
Comment (BytesGalore) [inline sys/stl-compatibility/condition_variable.cpp]: hmm, @josephnoir that's really a good question as [1] and [2] and others say to comment C++ like you did

[1] http://en.cppreference.com/w/cpp/comment
[2] http://en.wikibooks.org/wiki/C++_Programming/Code/Style_Conventions/Comments

**Comment by josephnoir** [comment]
It looks like the Travis error originate somewhere else. Am I missing something?

**Comment by OlegHahm** [comment]
Let's kick it and see.

**Comment by josephnoir** [comment]
Thanks. Travis seems to be happy now!
Next question: what do I need to change to comply to the RIOT guidelines? Probably more important: do you want this in RIOT and are you OK with the "hacks"? (i.e. the cpu stats thing in the sched.h)

File: core/include/sched.h [hunk 4]
```diff
@@ -188,6 +188,25 @@ void sched_register_cb(void (*callback)(uint32_t, uint32_t));
 
 #endif
 
+#ifdef BOARD_NATIVE
```
Comment (OlegHahm) [inline core/include/sched.h]: I would prefer to include a native header here that includes the following code block instead of having it directly part of `sched.h`.

File: cpu/stm32f4/startup.c [hunk 5]
```diff
@@ -41,6 +41,11 @@ extern void kernel_init(void);
 extern void __libc_init_array(void);
 
 /**
+ * Required by g++ cross compiler
+ */
+void* __dso_handle;
```
Comment (OlegHahm) [inline cpu/stm32f4/startup.c]: What's this and isn't required for other platforms as well?

And it should be `void *__dso_handle:`

**Comment by OlegHahm** [comment]
> Next question: what do I need to change to comply to the RIOT guidelines?

I think we will have to setup C++ specific guidelines first.

>  do you want this in RIOT and are you OK with the "hacks"?

For the hacks: see my comment and try to mark and isolate all hacks as much as possible. For the general question if we want this: I'm a complete C++ noob, but this looks very promising and helpful to me, so I would say: YES!

File: sys/stl-compatibility/Makefile [hunk 6]
```diff
@@ -0,0 +1,5 @@
+
```
Comment (BytesGalore) [inline sys/stl-compatibility/Makefile]: please remove the newline

File: sys/stl-compatibility/include/riot/chrono.hpp [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+#include <cstdio>
+#include <algorithm>
+
+#include "time.h"
+#include "vtimer.h"
+
+namespace riot {
+
+namespace {
+constexpr uint32_t microsecs_in_sec = 1000000;
```
Comment (BytesGalore) [inline sys/stl-compatibility/include/riot/chrono.hpp]: I think your comments on the class members/functions need some love ;)

File: tests/stl-compatibility_mutex/main.cpp [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+#include "riot/mutex.hpp"
+#include "riot/thread.hpp"
+#include "riot/condition_variable.hpp"
+
+using namespace std;
+using namespace riot;
+
+/* http://en.cppreference.com/w/cpp/thread/mutex */
+int main() {
+  printf("\n************ C++ mutex test ***********\n");
```
Comment (BytesGalore) [inline tests/stl-compatibility_mutex/main.cpp]: if you replace `printf()` with `puts()` in lines with only static text we can save some bytes in the codeimage

**Comment by BytesGalore** [comment]
@josephnoir it would be nice if you could migrate your tests to unittests,  (your tests are shaped like unittests anyway :+1: )

File: cpu/stm32f4/startup.c (hunk 5 above)
Comment (BytesGalore) [inline cpu/stm32f4/startup.c]: @OlegHahm the `void *__dso_handle;` is needed for destructors to provide the pointer to a handle for the DSO (Dynamic Shared Object) [1]. Unfortunately  I have no further information at hand now.

> isn't required for other platforms as well?

from my current state of knowledge yup, all platforms that are supposed to use the RIOT-STL wrapper will need this. (At least the ones compiled with `g++`)

[1] http://wiki.osdev.org/C%2B%2B#GCC

File: sys/stl-compatibility/include/riot/mutex.hpp [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+constexpr try_to_lock_t try_to_lock = try_to_lock_t();
+constexpr adopt_lock_t adopt_lock = adopt_lock_t();
+
+template <class Mutex>
+class lock_guard {
+
+ public:
+  using mutex_type = Mutex;
+
+  inline explicit lock_guard(mutex_type& mtx) : m_mtx(mtx) { m_mtx.lock(); }
```
Comment (BytesGalore) [inline sys/stl-compatibility/include/riot/mutex.hpp]: why `explicit`? Is there a constructor that may implicitly convert the passed parameter to another type silently?

File: sys/stl-compatibility/include/riot/mutex.hpp [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+  inline unique_lock(mutex_type& mtx, adopt_lock_t)
+      : m_mtx{&mtx}, m_owns{true} {}
+  inline ~unique_lock() {
+    if (m_owns) {
+      m_mtx->unlock();
+    }
+  }
+  inline unique_lock(unique_lock&& lock) noexcept : m_mtx{lock.m_mtx},
+                                                    m_owns{lock.m_owns} {
+    lock.m_mtx = nullptr;
```
Comment (BytesGalore) [inline sys/stl-compatibility/include/riot/mutex.hpp]: @josephnoir don't get it, how is the ownership of the lock transferred to this one?

File: core/include/sched.h (hunk 4 above)
Comment (josephnoir) [inline core/include/sched.h]: Would you include the header on a condition or include it anyways and have the `#ifdef` inside the header. Should it be placed in the core folder as well?

File: cpu/stm32f4/startup.c (hunk 5 above)
Comment (josephnoir) [inline cpu/stm32f4/startup.c]: This is probably required for other platforms as well. I have problems explaining why and when exactly it is needed, but my compiler complains otherwise. I will try to find out more ...

File: sys/stl-compatibility/include/riot/mutex.hpp (hunk 9 above)
Comment (josephnoir) [inline sys/stl-compatibility/include/riot/mutex.hpp]: Mainly because the [standard](http://en.cppreference.com/w/cpp/thread/lock_guard/lock_guard) says so.

Comment (BytesGalore) [inline sys/stl-compatibility/include/riot/mutex.hpp]: ok

File: sys/stl-compatibility/include/riot/mutex.hpp (hunk 10 above)
Comment (josephnoir) [inline sys/stl-compatibility/include/riot/mutex.hpp]: You take the configuration from the passed lock and reset its parameters.The [reference](http://en.cppreference.com/w/cpp/thread/unique_lock/unique_lock) uses the following words: `Move constructor. Initializes the unique_lock with the contents of other. Leaves other with no associated mutex.`

But  I am not sure I understand the question.

Comment (BytesGalore) [inline sys/stl-compatibility/include/riot/mutex.hpp]: ahh I just overlooked the member assignment of the constructor. :eyeglasses:

**Comment by BytesGalore** [comment]
@josephnoir just as you mentioned, it will be very tricky to get your c++ tests into the RIOT C unittest framework. 
Basically I'm happy with your tests just as they are. So strike my comment on this.

File: sys/stl-compatibility/condition_variable.cpp (hunk 1 above)
Comment (josephnoir) [inline sys/stl-compatibility/condition_variable.cpp]: [This](https://github.com/actor-framework/actor-framework/blob/master/HACKING.md) is the coding style we use for C++ in CAF. We also have a file to use with clangformat.

In RIOT it would probably be preferable to use astyle for C++ as well?

File: core/include/sched.h (hunk 4 above)
Comment (OlegHahm) [inline core/include/sched.h]: You could put it into `cpu.h` then you'll probably won't have to change anything here.

File: Makefile.dep [hunk 11]
```diff
@@ -130,3 +130,9 @@ ifneq (,$(filter nhdp,$(USEMODULE)))
 	USEMODULE += oonf_common
 	USEMODULE += oonf_rfc5444
 endif
+
+ifneq (,$(filter stl-compatibility,$(USEMODULE)))
```
Comment (Kijewski) [inline Makefile.dep]: Better call it cpp11-compat or anything similar. → http://stackoverflow.com/q/5205491/416224

File: core/include/sched.h (hunk 4 above)
Comment (josephnoir) [inline core/include/sched.h]: Simply moving the code to `cpu.h` does not work. Maybe the header is included for different modules?

File: Makefile.dep (hunk 11 above)
Comment (josephnoir) [inline Makefile.dep]: Sure, sounds plausible.

File: sys/stl-compatibility/include/riot/chrono.hpp (hunk 7 above)
Comment (josephnoir) [inline sys/stl-compatibility/include/riot/chrono.hpp]: I added some comments for the `chrono.hpp` as it is not compliant. As for the other headers, I can add an additional link to the specific entry on cppreference. I believe there is not much to gain, as either my comments are so simple or colse to a copy of the reference.

File: core/include/sched.h (hunk 4 above)
Comment (BytesGalore) [inline core/include/sched.h]: I think the problem is even more complicated, since the `pthread.h` from the host system requires the defines/types in a provided `sched.h`. So they must be available in the header, directly like currently or indirectly by include a further header.

Comment (josephnoir) [inline core/include/sched.h]: So, a new file, i.e., `native_sched,h`, with the code in it and the according include in `sched.h`.

Comment (BytesGalore) [inline core/include/sched.h]: yes, a separate header with the compatibility additions is highly preferable :+1: .

**Comment by josephnoir** [comment]
Ok. So what is missing now is an explanation for the `__dso_handle`?

**Comment by OlegHahm** [comment]
What do you mean: Figuring out why this is needed or documentation?

**Comment by josephnoir** [comment]
The code does not compile without it. So, I guess both steps are required. First find out why it wasn't required previously and then write an appropriate comment in case the question comes up in the future.

File: sys/Makefile [hunk 12]
```diff
@@ -17,8 +17,8 @@ ifneq (,$(filter nomac,$(USEMODULE)))
     DIRS += net/link_layer/nomac
 endif
 ifneq (,$(filter transport_layer,$(USEMODULE)))
-	USEMODULE += udp
-	USEMODULE += tcp
+  USEMODULE += udp
+  USEMODULE += tcp
```
Comment (BytesGalore) [inline sys/Makefile]: This is unrelated for the RP, @OlegHahm but why this is done twice?
The mmodules are also included in `Makefile.dep` [1]

[1] https://github.com/RIOT-OS/RIOT/blob/master/Makefile.dep#L12

Comment (BytesGalore) [inline sys/Makefile]: From my side we can keep the edits if you change it to 2->4 white-spaces (also  in line 48 below), or just revert the changes for the lines.

**Comment by BytesGalore** [comment]
Beside the nitpick with the whitespaces ACK from my side.
I vote to enhance the documentation for `__dso_handle` later, since it is needed anyways for shared objects/destructors.

File: sys/Makefile (hunk 12 above)
Comment (OlegHahm) [inline sys/Makefile]: @BytesGalore, looks weird. I think this is not on purpose.

**Comment by BytesGalore** [comment]
travis is happy, my ACK holds -> and go C++11 !!!!

**Comment by josephnoir** [comment]
Yeah :)

**Comment by BytesGalore** [comment]
congratz man, this is a giant leap :)

**Comment by OlegHahm** [comment]
:tada: :confetti_ball:
